File: CS/dxSample/Custom Combo Box/CustomComboBoxEdit.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Globalization;
using System.Linq;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Controls;

namespace dxSample {
    [ToolboxItem(true)]
    public class CustomComboBoxEdit : ComboBoxEdit {
        static CustomComboBoxEdit() {
            CustomRIComboBox.Register();
        }
        public override string EditorTypeName {
            get {
                return CustomRIComboBox.EditorName;
            }
        }
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new CustomRIComboBox Properties {
            get {
                return base.Properties as CustomRIComboBox;
            }
        }
        public CustomComboBoxEdit() {
        }
        protected override int FindItem(string text, bool autoComplete, int startIndex) {
            string resText = text;
            if (!Properties.CaseSensitiveSearch) {
                resText = resText.ToLower();
            }
            int filterStr = FindStringEx(resText, startIndex);
            return filterStr;
        }
        public override string AutoSearchText {
            get {
                string s = base.AutoSearchText;
                if (String.IsNullOrWhiteSpace(s) || string.IsNullOrEmpty(s)) {
                    return s;
                }
                return s.Substring(s.Length - 1, 1);
            }
            set {
                base.AutoSearchText = value;
            }
        }
        public int ItemCount {
            get {
                return Properties.Items.Count;
            }
        }
        private int FindStringEx(string s, int startIndex) {
            if (s == null) {
                return -1;
            }
            ComboBoxItemCollection itemsCollection = Properties.Items;
            if ((itemsCollection == null) || (itemsCollection.Count == 0)) {
                return -1;
            }
            if ((startIndex < -1) || (startIndex >= itemsCollection.Count)) {
                throw new ArgumentOutOfRangeException("startIndex");
            }
            return FindStringInternal(s, Properties.Items, startIndex, false);
        }
        internal int FindStringInternal(string str, IList items, int startIndex, bool exact) {
            return FindStringInternal(str, items, startIndex, exact, true);
        }
        private int _LastFindIndex = -1;
        public int LastFindIndex {
            get {
                return _LastFindIndex;
            }
            set {
                _LastFindIndex = value;
            }
        }
        internal int FindStringInternal(string str, IList items, int startIndex, bool exact, bool ignorecase) {
            startIndex = GetStartIndex();
            if ((str != null) && (items != null)) {
                if ((startIndex < -1) || (startIndex >= items.Count)) {
                    return -1;
                }
                bool flag = false;
                int length = str.Length;
                int num2 = 0;
                for (int i = (startIndex + 1) % items.Count; num2 < items.Count; i = (i + 1) % items.Count) {
                    num2++;
                    string item = items[i].ToString();
                    flag = string.Compare(str, 0, item, 0, length, ignorecase, CultureInfo.CurrentCulture) == 0;
                    if (flag) {
                        return i;
                    }
                }
            }
            return -1;
        }
        private int GetStartIndex() {
            if (IsPopupOpen) {
                return  PopupForm.SelectedItemIndex;
            }
            return SelectedIndex;
        }
    }
}

File: VB/dxSample/Custom Combo Box/CustomComboBoxEdit.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Globalization
Imports System.Linq
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Controls

Namespace dxSample
	<ToolboxItem(True)>
	Public Class CustomComboBoxEdit
		Inherits ComboBoxEdit

		Shared Sub New()
			CustomRIComboBox.Register()
		End Sub
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return CustomRIComboBox.EditorName
			End Get
		End Property
		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)>
		Public Shadows ReadOnly Property Properties() As CustomRIComboBox
			Get
				Return TryCast(MyBase.Properties, CustomRIComboBox)
			End Get
		End Property
		Public Sub New()
		End Sub
		Protected Overrides Function FindItem(ByVal text As String, ByVal autoComplete As Boolean, ByVal startIndex As Integer) As Integer
			Dim resText As String = text
			If Not Properties.CaseSensitiveSearch Then
				resText = resText.ToLower()
			End If
			Dim filterStr As Integer = FindStringEx(resText, startIndex)
			Return filterStr
		End Function
		Public Overrides Property AutoSearchText() As String
			Get
				Dim s As String = MyBase.AutoSearchText
				If String.IsNullOrWhiteSpace(s) OrElse String.IsNullOrEmpty(s) Then
					Return s
				End If
				Return s.Substring(s.Length - 1, 1)
			End Get
			Set(ByVal value As String)
				MyBase.AutoSearchText = value
			End Set
		End Property
		Public ReadOnly Property ItemCount() As Integer
			Get
				Return Properties.Items.Count
			End Get
		End Property
		Private Function FindStringEx(ByVal s As String, ByVal startIndex As Integer) As Integer
			If s Is Nothing Then
				Return -1
			End If
			Dim itemsCollection As ComboBoxItemCollection = Properties.Items
			If (itemsCollection Is Nothing) OrElse (itemsCollection.Count = 0) Then
				Return -1
			End If
			If (startIndex < -1) OrElse (startIndex >= itemsCollection.Count) Then
				Throw New ArgumentOutOfRangeException("startIndex")
			End If
			Return FindStringInternal(s, Properties.Items, startIndex, False)
		End Function
		Friend Function FindStringInternal(ByVal str As String, ByVal items As IList, ByVal startIndex As Integer, ByVal exact As Boolean) As Integer
			Return FindStringInternal(str, items, startIndex, exact, True)
		End Function
		Private _LastFindIndex As Integer = -1
		Public Property LastFindIndex() As Integer
			Get
				Return _LastFindIndex
			End Get
			Set(ByVal value As Integer)
				_LastFindIndex = value
			End Set
		End Property
		Friend Function FindStringInternal(ByVal str As String, ByVal items As IList, ByVal startIndex As Integer, ByVal exact As Boolean, ByVal ignorecase As Boolean) As Integer
			startIndex = GetStartIndex()
			If (str IsNot Nothing) AndAlso (items IsNot Nothing) Then
				If (startIndex < -1) OrElse (startIndex >= items.Count) Then
					Return -1
				End If
				Dim flag As Boolean = False
				Dim length As Integer = str.Length
				Dim num2 As Integer = 0
				Dim i As Integer = (startIndex + 1) Mod items.Count
				Do While num2 < items.Count
					num2 += 1
					Dim item As String = items(i).ToString()
					flag = String.Compare(str, 0, item, 0, length, ignorecase, CultureInfo.CurrentCulture) = 0
					If flag Then
						Return i
					End If
					i = (i + 1) Mod items.Count
				Loop
			End If
			Return -1
		End Function
		Private Function GetStartIndex() As Integer
			If IsPopupOpen Then
				Return PopupForm.SelectedItemIndex
			End If
			Return SelectedIndex
		End Function
	End Class
End Namespace
